Parser/Util/PSIPrint.cs

```diff
@@ -20,12 +20,40 @@ public override StringBuilder Visit (NBlock b)
             NWrite ($"{g.Select (a => a.Name).ToCSV ()} : {g.Key};");
          N--;
       }
+
+      if (d.Methods.Length > 0) {
+         NWrite ("");
+         N++;
+         foreach (var method in d.Methods) {
+            Visit (method);
+            Write (";\n");
+         }
+         N--;
+      }
+
       return S;
    }
 
    public override StringBuilder Visit (NVarDecl d)
       => NWrite ($"{d.Name} : {d.Type}");
 
+   public override StringBuilder Visit (NFnDecl d) {
+      NWrite ($"function {Declaration (d)} : {d.Type};");
+      return Visit (d.Block);
+   }
+
+   public override StringBuilder Visit (NProcDecl d) {
+      NWrite ($"procedure {Declaration (d)};"); 
+      return Visit (d.Block);
+   }
+
+   string Declaration (NMethodDecl d) {
+      return $"{d.Name.Text} ({Params (d.Params)})";
+
+      string Params (NVarDecl[][] VarsList) => VarsList.Select (Param).ToCSV ("; ");

```

Comment:
While this matches the grammar, it would be simpler downstream to have a single flat array of NVarDecl rather than a 2-dimensional array. When printing the code, we could use GroupBy to combine vars with same types together.